apps/api/src/tests/services/application.ts

```diff
@@ -0,0 +1,197 @@
+			.start();
+
+		const connectionString = container.getConnectionUri();
+		db = startDb(connectionString);
+
+		await initMigration(db);
+
+		applicationService = service(db);
+		// TODO: create file with seed data for postgres to test read actions
+		// Not yet supported in Drizzle Kit:
+		// https://orm.drizzle.team/docs/kit-seed-data
+	});
+
+	it('should create applications with status DRAFT and submitted user_id', async () => {
+		const application = await applicationService.createApplication({ user_id });
+
+		assert.notEqual(application, null);
+		assert.strictEqual(application?.user_id, user_id);
+		assert.strictEqual(application?.state, ApplicationStates.DRAFT);
+	});
+
+	it('should get applications requested by id, with application_contents', async () => {
+		const applicationRecords = await applicationService.listApplications({ user_id });
+
+		assert.ok(Array.isArray(applicationRecords));
+
+		const { id } = applicationRecords[0];
+
+		const requestedApplication = await applicationService.getApplicationById({ id });
+
+		assert.notEqual(requestedApplication, null);
+		assert.strictEqual(requestedApplication?.id, id);
+		assert.strictEqual(requestedApplication?.id, requestedApplication?.contents?.application_id);
+	});
+
+	it('should list applications filtered by user_id', async () => {
+		await applicationService.createApplication({ user_id });
+		await applicationService.createApplication({ user_id });
+
+		const applicationRecords = await applicationService.listApplications({ user_id });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+	});
+
+	it('should list applications filtered by state', async () => {
+		const applicationRecords = await applicationService.listApplications({ state: ApplicationStates.DRAFT });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+	});
+
+	it('should allow sorting records by created_at', async () => {
+		const applicationRecords = await applicationService.listApplications({ sort: 'created_at' });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+
+		const date1 = applicationRecords[0].createdAt.valueOf();
+		const date2 = applicationRecords[1].createdAt.valueOf();
+		const date3 = applicationRecords[2].createdAt.valueOf();
+
+		assert.ok(date1 < date2);
+		assert.ok(date2 < date3);
+	});
+
+	it('should allow sorting records by updated_at', async () => {
+		const applicationRecords = await applicationService.listApplications({ user_id });
+
+		assert.ok(Array.isArray(applicationRecords));
+		assert.strictEqual(applicationRecords.length, 3);
+
+		const { id: zeroRecordId } = applicationRecords[0];
+		const { id: firstRecordId } = applicationRecords[1];
+		const { id: secondRecordId } = applicationRecords[2];
+
+		// State values are used in the next test so first record remains in 'Draft', this is just populating `updatedAt`
```

Comment:
TODO: Tests need to be independent or grouped together as a single test with sub tests.

Reply:
Yes definitely. I parked some of the test setup tasks because Drizzle seeding is not available yet (see comment at line 56 of this file).

The tests are written sequentially (so that Test B depends on Test A etc.) but this is not ideal (or sustainable), and I intended to update this before we went much further.

I opened the PR at this state because it was functionally complete, and I wanted to take a step back to assess how we should approach test setup with the team's input.

Reply:
Tests are now grouped together and setup functions enable independent execution